lib/WebService/Async/CustomerIO/RateLimiter.pm

```diff
@@ -0,0 +1,94 @@
+package WebService::Async::CustomerIO::RateLimiter;
+
+use strict;
+use warnings;
+
+use Carp qw();
+use Future;
+use mro;
+
+use parent qw(IO::Async::Notifier);
+
+=head1 NAME
+WebService::Async::CustomerIO::RateLimitter - This class provide possibility to limit amount
+of request in time interval
+
+=head1 SYNOPSIS
+
+
+=head1 DESCRIPTION
+
+=cut
+sub configure {
+    my ($self, %args) = @_;
+    for my $k (qw(limit interval)) {
+        die "Missing required argument: $k" unless exists $args{$k};
+        die "Invalid value for $k: $args{$k}" unless int($args{$k}) > 0;
+        $self->{$k} = delete $args{$k} if exists $args{$k};
+    }
+
+    $self->{queue} = [];
+    $self->{counter} = 0;
+
+    $self->next::method(%args);
+}
+
+=head2 interval
+=cut
+sub interval {shift->{interval}}
+
+=head2 limit
+=cut
+sub limit {shift->{limit}}
+
+=head2 acquire
+
+Method checks availability for free slot.
+It returns future, when slot will be available, then future will be resolved.
+
+=cut
+sub acquire {
+    my ($self) = @_;
+
+    $self->_start_timer;
+    return Future->done if ++$self->{counter} <= $self->limit;
+
+    my $current = $self->_current_queue;
+    $current->{counter}++;
+    return $current->{future};
+}
+
+sub _current_queue {
+    my ($self) = @_;
+
+    # +1 for getting correct position for edge cases like: limit 2, counter 4, should be 0
+    my $pos = int(($self->{counter} - ($self->limit + 1)) / $self->limit);
+
+    $self->{queue}[$pos] //= {future => Future->new, counter=> 0};
```

Comment:
This won't work - you can't wait for a plain `Future` instance, in general I'd recommend *never* calling `Future->new` directly. Instead, `$loop->new_future`.